CAF::Lock does an unlock() in its DESTROY method

During the installation of a rhel5 system, we get the following error:

[ERROR] cannot release lock file: /var/lock/quattor/ncm-ncd

This occurs a lot (at least once for every component it seems, maybe
more. Nothing goes wrong though.

Gabor tracked this down to the following:

> The Sys::Syslog Perl module that comes with RH5 does not use syslog()
> from libc, but tries to re-implement it. When it detects that there's
> no syslog service running, it falls back to printing messages to the
> console. In that mode, it does a fork() every time a message is
> printed. When such a child process exists, the Perl interpreter calls
> the DESTROY method of the NCM lock object, which removes the lock
> file. The next time syslog() is called, or when the main process
> finishes, it tries to release the lock again, but it is no longer
> there, therefore the error message.
> 
> CAF::Lock doing an unlock() in its DESTROY method is conceptually
> wrong, since it means applications using it and doing a fork() may
> silently lose any locks they hold while thinking they still hold
> it. Leaving stale lock files around if the application fails to call
> unlock() explicitly would be a better failure mode.

--- a/src/main/perl/Lock.pm
+++ b/src/main/perl/Lock.pm
@@ -304,7 +304,10 @@ application instance.
 
 sub DESTROY {
   my $self = shift;
-  $self->unlock() if ($self->{LOCK_SET});
+  # We unlock only on the process that owns the lock.  Otherwise this
+  # might be a forked process that is exiting and shouldn't sweep
+  # under its parent's feet.
+  $self->unlock() if $self->{LOCK_SET} && $self->get_lock_pid() == $$;
 }
 
 

--- a/src/test/perl/lock-fork.t
+++ b/src/test/perl/lock-fork.t
@@ -1,0 +1,33 @@
+# -*- mode: cperl -*-
+use strict;
+use warnings;
+use Test::More;
+use CAF::Lock qw(FORCE_IF_STALE FORCE_ALWAYS);
+use Test::MockObject::Extends;
+
+use constant LOCK_TEST_DIR => "target/tests";
+use constant LOCK_TEST => LOCK_TEST_DIR . "/lock-fork";
+
+my $pid;
+
+my $mock = Test::MockObject::Extends->new("CAF::Lock");
+
+$mock->mock('unlock', 1);
+
+mkdir(LOCK_TEST_DIR);
+unlink(LOCK_TEST);
+
+my $lock=CAF::Lock->new(LOCK_TEST);
+
+$lock->set_lock();
+
+$lock=CAF::Lock->new(LOCK_TEST);
+
+$$++;
+
+$lock = undef;
+
+$mock->next_call();
+is($mock->next_call(), undef, "Lock is not released after forking");
+
+done_testing();

